core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/recipe/OrchestratorRecipeExecutor.java

```diff
@@ -176,30 +177,42 @@ private String getRecipeTimeoutErrorMessage(CloudbreakOrchestratorTimeoutExcepti
                 " minute(s), please check your recipe(s) and recipe logs on the machines under /var/log/recipes! Reason:" + timeoutException.getMessage();
     }
 
-    private String getRecipeExecutionFaiureMessage(Stack stack, CloudbreakOrchestratorException exception) {
+    private String getRecipeExecutionFailureMessage(Stack stack, CloudbreakOrchestratorException exception) {
+        LOGGER.info("Getting execution failure message in stack {} for exception", stack.getId(), exception);
         if (!recipeExecutionFailureCollector.canProcessExecutionFailure(exception)) {
             return exception.getMessage();
         }
         List<RecipeFailure> failures = recipeExecutionFailureCollector.collectErrors(exception);
         Set<InstanceMetaData> instanceMetaData = instanceMetaDataService.getAllInstanceMetadataByStackId(stack.getId());
 
-        String message = failures.stream().map(failure -> {
-            InstanceMetaData metadata = recipeExecutionFailureCollector.getInstanceMetadataByHost(instanceMetaData, failure.getHost()).get();
-            return new StringBuilder("[Recipe: '")
-                    .append(failure.getRecipeName())
-                    .append("' - \n")
-                    .append("Hostgroup: '")
-                    .append(metadata.getInstanceGroup().getGroupName())
-                    .append("' - \n")
-                    .append("Instance: '")
-                    .append(metadata.getDiscoveryFQDN())
-                    .append(']')
-                    .toString();
-        }).collect(Collectors.joining("\n ---------------------------------------------- \n")
-        );
+        String message = failures.stream()
+                .map(failure -> getSingleRecipeExecutionFailureMessage(instanceMetaData, failure))
+                .collect(Collectors.joining("\n ---------------------------------------------- \n"));
         return new StringBuilder("Failed to execute recipe(s): \n").append(message).toString();
     }
 
+    @VisibleForTesting
+    String getSingleRecipeExecutionFailureMessage(Set<InstanceMetaData> instanceMetaData, RecipeFailure failure) {
```

Comment:
I would prefer not using this `@VisibleForTesting` but unit test it through public method and make this private

Reply:
This is called from another private method, that handles error from 4 different public methods, this is why I have done it this way. To eliminate this annotation a better solution would be to extract this logic to its own class (or into RecipeExecutionFailureCollector) and test it instead, I can do it if you think that would be better.

Reply:
+1 on reducing the complexity of this class by moving out the failure message generation logic into a separated component

Reply:
i think testing only one public method's error handling would be enough to test this method